File: Day 4/Task 2/Task 2.cpp
#include <iostream>
#include <fstream>
#include <regex>
#include <string>
#include <utility>


//check if one of the two ranges contains the other
bool OverlappingRanges(const std::pair<int, int>& firstInterval, const std::pair<int, int>& secondInterval) {
	if (firstInterval.second >= secondInterval.first && firstInterval.first <= secondInterval.second)
		return true;
	else if (secondInterval.second >= firstInterval.first && secondInterval.first <= firstInterval.second)
		return true;

	return false;
}

int main()
{
	std::pair<int, int> firstInterval, secondInterval;
	std::ifstream fhandle("day_4_1.txt");

	std::string line;
	std::regex pattern("([0-9]+)\-([0-9]+),([0-9]+)\-([0-9]+)");
	std::smatch ranges; // match object for std::string containing the matched pattern and all captures form the pattern

	int numOfOverlapping = 0;
	while (std::getline(fhandle, line)) {
		if (std::regex_search(line, ranges, pattern)) {

			firstInterval = std::make_pair(std::stoi(ranges[1]), std::stoi(ranges[2]));
			secondInterval = std::make_pair(std::stoi(ranges[3]), std::stoi(ranges[4]));

			if (OverlappingRanges(firstInterval, secondInterval))
				numOfOverlapping++;

		}
	}
	std::cout << "Number of assignment pairs where one range fully contains the other: " << numOfOverlapping;



}

File: Day 4/Task 2/Task_2.py
import re

def overlapping_ranges(firstInterval, secondInterval):
    if secondInterval[1] >= firstInterval[0] and secondInterval[0] <= firstInterval[1]:
        return True
    elif firstInterval[1] >= secondInterval[0] and firstInterval[0] <= secondInterval[1]:
        return  True
    return False


num_of_overlapping = 0

with open("day_4_1.txt") as fhandle:
    for line in fhandle:
        line = line.rstrip()
        ranges = re.findall("([0-9]+)\-([0-9]+)", line)
        if ranges is not None:
            ranges = [(int(x), int(y)) for x, y in ranges]

            if overlapping_ranges(ranges[0], ranges[1]):
                num_of_overlapping += 1

print('Number of assignment pairs where one range fully contains the other: ', num_of_overlapping)


